File: todo-web-app/frontend/src/types/notification.ts
export interface Notification {
    id: number;
    user_id: string;
    title: string;
    message: string;
    task_id?: number;
    type?: "info" | "success" | "warning" | "error";
    category?: "system" | "reminder" | "task" | "achievement";
    link?: string | null;
    is_read: boolean;
    created_at: string;
}

export interface NotificationListResponse {
    items: Notification[];
    unread_count: number;
}

File: todo-web-app/backend/src/models/notification.py
from datetime import UTC, datetime
from sqlmodel import Field, SQLModel, Relationship

def utc_now() -> datetime:
    """Return current UTC datetime as naive datetime for PostgreSQL compatibility."""
    return datetime.now(UTC).replace(tzinfo=None)

class NotificationBase(SQLModel):
    user_id: str = Field(index=True)
    task_id: int | None = Field(default=None, foreign_key="task.id", nullable=True)
    title: str
    message: str
    type: str = Field(default="info", max_length=20) # info, success, warning, error
    category: str = Field(default="system", max_length=20) # system, reminder, task, achievement
    link: str | None = Field(default=None, max_length=500)
    is_read: bool = Field(default=False, index=True)
    created_at: datetime = Field(default_factory=utc_now)

class Notification(NotificationBase, table=True):
    id: int | None = Field(default=None, primary_key=True)

    task: "Task" = Relationship(
        back_populates="notifications",
        sa_relationship_kwargs={"lazy": "selectin"}
    )

class NotificationCreate(SQLModel):
    title: str
    message: str
    task_id: int | None = None

class NotificationUpdate(SQLModel):
    is_read: bool | None = None
